common/flatpak-run.c

```diff
@@ -1691,6 +1692,180 @@ flatpak_run_allocate_id (int *lock_fd_out)
+    }
+  g_string_append_c (s, '/');
+
+  return g_string_free (s, FALSE);
+}
+
+static void
+get_dconf_data (const char  *app_id,
+                const char **settings,
+                char **defaults,
+                gsize *defaults_size,
+                char **locks,
+                gsize *locks_size)
+{
+  DConfClient *client = NULL;
+  g_autoptr(GKeyFile) keyfile = NULL;
+  g_autofree char *path = NULL;
+  g_autoptr(GError) error = NULL;
+  g_autoptr(GString) contents = NULL;
+  g_autofree char *prefix = NULL;
+  g_autofree char *dir = NULL;
+
+  prefix = dconf_path_for_app_id (app_id);
+
+  client = dconf_client_new ();
+  keyfile = g_key_file_new ();
+
+  contents = g_string_new ("");
+
+  g_debug ("Add defaults in dir %s", prefix);
+  add_dconf_dir_to_keyfile (keyfile, client, prefix);
+
+  g_debug ("Add locks in dir %s", prefix);
+  add_dconf_locks_to_list (contents, client, prefix);
+
+  if (settings)
+    {
+      int i;
+      for (i = 0; settings[i]; i++)
+        {
+          if (dconf_is_dir (settings[i], NULL))
+            {
+              g_debug ("Add defaults in dir %s", settings[i]);
+              add_dconf_dir_to_keyfile (keyfile, client, settings[i]);
+
+              g_debug ("Add locks in dir %s", settings[i]);
+              add_dconf_locks_to_list (contents, client, settings[i]);
+            }
+          else if (dconf_is_key (settings[i], NULL))
+            {
+              g_debug ("Add individual key %s", settings[i]);
+              add_dconf_key_to_keyfile (keyfile, client, settings[i]);
+            }
+          else
+            {
+              g_warning ("Ignoring settings path '%s': neither dir nor key", settings[i]);
+            }
+        }
+    }
+
+  *defaults = g_key_file_to_data (keyfile, defaults_size, NULL);
+  *locks = g_strdup (contents->str);
+  *locks_size = contents->len;
+
+  g_object_unref (client);
+}
+
+static gboolean
+flatpak_run_add_dconf_args (FlatpakBwrap  *bwrap,
+                            const char    *app_id,
+                            GKeyFile      *metakey,
+                            GError       **error)
+{
+  g_auto(GStrv) settings = NULL;
+  g_autofree char *defaults = NULL;
+  g_autofree char *locks = NULL;
+  gsize defaults_size;
+  gsize locks_size;
+
+  if (metakey)
+    settings = g_key_file_get_string_list (metakey,
+                                           FLATPAK_METADATA_GROUP_DCONF,
+                                           FLATPAK_METADATA_KEY_DCONF_PATHS,
+                                           NULL, NULL);
+ 
+  get_dconf_data (app_id, (const char **)settings,
+                           &defaults, &defaults_size,
+                           &locks, &locks_size);
+
+  if (//defaults_size != 0 &&
+      !flatpak_bwrap_add_args_data (bwrap,
+                                    "dconf-defaults",
+                                    defaults, defaults_size,
+                                    "/etc/glib-2.0/settings/defaults",
```

Comment:
This might accidentally conflict with a file that is already existing in /etc. Well, rather it would be in /usr/etc, but we might have made a symlink there.

To avoid this we need to explicitly ignore this from the runtime in  flatpak_run_setup_base_argv() among with the other /etc stuff. Do we really want to disallow the runtime shipping an /etc/glib-2.0 though? Seems like that could be useful.

Reply:
waiting with changes here until we reach consensus on the glib side